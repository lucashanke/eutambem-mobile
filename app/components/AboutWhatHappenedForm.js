import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { View } from 'react-native';

import appStyles from '../styles';

import { Button, Picker, DatePicker, TextInput, CheckBoxGroup, PlaceInput } from './common';

import { MAX_TEXT_INPUT_LENGTH } from '../constants';

import { formWrapper, loading } from './hoc';

export class AboutWhatHappenedForm extends Component {
  constructor(props) {
    super(props);
    this.state = {
      formData: {
        establishment: {},
        harassmentType: '',
        description: '',
        date: null,
        followupActions: [],
        wouldRecommend: '',
        advice: '',
      },
    };
  }

  updateFormDataValue = (key, value) => {
    const { formData } = this.state;
    if (key in formData) {
      formData[key] = value;
      this.setState({ formData });
    } else {
      console.error('Trying to update key that is not present in formData.');
    }
  }

  render() {
    return (
      <View>
        <PlaceInput
          placeholder="Empresa em que ocorreu"
          value={this.state.formData.establishment.label}
          onValueChange={item => this.updateFormDataValue('establishment', { value: item.id, label: item.label })}
          types={['establishment']}
          navigation={this.props.navigation}
          required
          placeScreenTitle="Informe a Empresa"
        />
        <Picker
          required
          placeholder="Tipo de assédio"
          value={this.state.formData.harassmentType}
          onValueChange={value => this.updateFormDataValue('harassmentType', value)}
          items={this.props.data.formOptions.harassment_type_options}
        />
        <TextInput
          multiline
          required
          placeholder="Conte-nos mais sobre o ocorrido"
          onChangeText={value => this.updateFormDataValue('description', value)}
          value={this.state.formData.description}
        />
        <DatePicker
          date={this.state.formData.date}
          required
          placeholder="Quando ocorreu"
          maxDate={new Date(Date.now())}
          onDateChange={date => this.updateFormDataValue('date', date)}
        />
        <CheckBoxGroup
          label="Você tomou alguma providência com relação ao ocorrido?"
          onItemToggle={values => this.updateFormDataValue('followupActions', values)}
          options={this.props.data.formOptions.followup_actions_options}
        />
        <Picker
          required
          value={this.state.formData.wouldRecommend}
          onValueChange={value => this.updateFormDataValue('wouldRecommend', value)}
          placeholder="Você recomendaria essa empresa depois do ocorrido?"
          items={this.props.data.formOptions.yes_no_optional_options}
        />
        <TextInput
          style={[appStyles.input, appStyles.multilineInput]}
          multiline
          autoGrow
          maxLength={MAX_TEXT_INPUT_LENGTH}
          placeholder="Qual seu conselho para os gestores?"
          onChangeText={advice => this.updateFormDataValue('advice', advice)}
          value={this.state.formData.advice}
        />
        <Button
          onPress={() => this.props.navigation.navigate('AboutYou', { formData: this.state.formData })}
          title="Prosseguir"
        />
      </View>
    );
  }
}

const valueLabelShape = PropTypes.arrayOf(PropTypes.shape({
  value: PropTypes.string,
  label: PropTypes.string,
})).isRequired;

AboutWhatHappenedForm.propTypes = {
  navigation: PropTypes.shape({
    navigate: PropTypes.func,
  }).isRequired,
  data: PropTypes.shape({
    formOptions: PropTypes.shape({
      harassment_type_options: valueLabelShape,
      followup_actions_options: valueLabelShape,
      yes_no_optional_options: valueLabelShape,
    }).isRequired,
  }).isRequired,
};

export default formWrapper(loading(AboutWhatHappenedForm));
