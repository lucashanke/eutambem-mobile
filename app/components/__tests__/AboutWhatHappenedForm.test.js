import 'react-native';
import React from 'react';
import { shallow } from 'enzyme';
import sinon from 'sinon';

import { AboutWhatHappenedForm } from '../AboutWhatHappenedForm';


export const FOLLOWUP_ACTIONS_OPTIONS = [{
  key: 'talkedToAccountable',
  label: 'Falei com algum Responsável (RH, Superior direto, etc)',
}, {
  key: 'gaveFeedback',
  label: 'Dei o feedback para o envolvido',
}, {
  key: 'policeReport',
  label: 'Fiz um Boletim de Ocorrência',
}];

export const HARASSMENT_TYPE_OPTIONS = [{
  value: 'physical',
  label: 'Assédio Físico',
}, {
  value: 'verbal',
  label: 'Assédio Verbal',
}, {
  value: 'homophobia',
  label: 'Homofobia',
}, {
  value: 'racism',
  label: 'Racismo',
}, {
  value: 'transfobia',
  label: 'Transfobia',
}, {
  value: 'other',
  label: 'Outro tipo de Assédio',
}];

export const YES_NO_OPTIONAL_OPTIONS = [{
  value: 'yes',
  label: 'Sim',
}, {
  value: 'no',
  label: 'Não',
}, {
  value: 'notInformed',
  label: 'Prefiro não responder',
}];

const props = {
  navigation: {
    navigate: sinon.spy(),
  },
  data: {
    formOptions: {
      followup_actions_options: FOLLOWUP_ACTIONS_OPTIONS,
      harassment_type_options: HARASSMENT_TYPE_OPTIONS,
      yes_no_optional_options: YES_NO_OPTIONAL_OPTIONS,
    },
  },
};

describe('AboutWhatHappenedForm', () => {
  it('renders correctly', () => {
    const wrapper = shallow(<AboutWhatHappenedForm {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('passes the form data to the AboutYou screen', () => {
    const wrapper = shallow(<AboutWhatHappenedForm {...props} />);
    const initialFormData = wrapper.state().formData;
    const initialFormDataValues = wrapper.instance().getFormDataValues();
    wrapper.instance().isValid = () => true;

    wrapper.setState({
      formData: {
        ...initialFormData,
        description: { value: 'something' },
      },
    });

    wrapper.find('Button').simulate('press');

    expect(props.navigation.navigate.calledWith('AboutYou', {
      formData: {
        ...initialFormDataValues,
        description: 'something',
      },
    })).toBeTruthy();
  });
});
