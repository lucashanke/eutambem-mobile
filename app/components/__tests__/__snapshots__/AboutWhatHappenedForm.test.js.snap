// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AboutWhatHappenedForm renders correctly 1`] = `
<ForwardRef(View)>
  <Input
    navigation={
      Object {
        "navigate": [Function],
      }
    }
    onValueChange={[Function]}
    placeScreenTitle="Informe a Empresa"
    placeholder="Empresa em que ocorreu"
    required={true}
    showValidation={false}
    types={
      Array [
        "establishment",
      ]
    }
  />
  <Input
    items={
      Array [
        Object {
          "label": "Assédio Físico",
          "value": "physical",
        },
        Object {
          "label": "Assédio Verbal",
          "value": "verbal",
        },
        Object {
          "label": "Homofobia",
          "value": "homophobia",
        },
        Object {
          "label": "Racismo",
          "value": "racism",
        },
        Object {
          "label": "Transfobia",
          "value": "transfobia",
        },
        Object {
          "label": "Outro tipo de Assédio",
          "value": "other",
        },
      ]
    }
    onValueChange={[Function]}
    placeholder="Tipo de assédio"
    required={true}
    showValidation={false}
    value=""
  />
  <Input
    multiline={true}
    onValueChange={[Function]}
    placeholder="Conte-nos mais sobre o ocorrido"
    required={true}
    showValidation={false}
    value=""
  />
  <Input
    maxDate={1970-01-18T14:58:09.340Z}
    onValueChange={[Function]}
    placeholder="Quando ocorreu"
    required={true}
    showValidation={false}
    value={null}
  />
  <Input
    label="Você tomou alguma providência com relação ao ocorrido?"
    onValueChange={[Function]}
    options={
      Array [
        Object {
          "key": "talkedToAccountable",
          "label": "Falei com algum Responsável (RH, Superior direto, etc)",
        },
        Object {
          "key": "gaveFeedback",
          "label": "Dei o feedback para o envolvido",
        },
        Object {
          "key": "policeReport",
          "label": "Fiz um Boletim de Ocorrência",
        },
      ]
    }
    required={false}
    showValidation={false}
    values={Array []}
  />
  <Input
    items={
      Array [
        Object {
          "label": "Sim",
          "value": "yes",
        },
        Object {
          "label": "Não",
          "value": "no",
        },
        Object {
          "label": "Prefiro não responder",
          "value": "notInformed",
        },
      ]
    }
    onValueChange={[Function]}
    placeholder="Você recomendaria essa empresa depois do ocorrido?"
    required={true}
    showValidation={false}
    value=""
  />
  <Input
    autoGrow={true}
    maxLength={350}
    multiline={true}
    onValueChange={[Function]}
    placeholder="Qual seu conselho para os gestores?"
    required={false}
    showValidation={false}
    style={
      Array [
        Object {
          "alignItems": "flex-start",
          "backgroundColor": "#fcfcfc",
          "borderRadius": 5,
          "borderWidth": 0,
          "height": 30,
          "margin": 0,
          "marginBottom": 10,
          "marginTop": 10,
          "padding": 5,
        },
        Object {
          "height": "auto",
          "minHeight": 30,
        },
      ]
    }
    value=""
  />
  <Button
    onPress={[Function]}
    title="Prosseguir"
    type="confirm"
  />
</ForwardRef(View)>
`;
